test/iptables/iptables_test.go

```diff
@@ -65,15 +65,47 @@ func singleTest(t *testing.T, test TestCase) {
 	}
 
 	// Run our side of the test.
-	if err := test.LocalAction(ip); err != nil {
-		t.Fatalf("LocalAction failed: %v", err)
-	}
-
-	// Wait for the final statement. This structure has the side effect
-	// that all container logs will appear within the individual test
-	// context.
-	if _, err := d.WaitForOutput(ctx, TerminalStatement, TestTimeout); err != nil {
-		t.Fatalf("test failed: %v", err)
+	localCh := make(chan error)
+	go func() {
+		if err := test.LocalAction(ip); err != nil {
+			localCh <- fmt.Errorf("LocalAction failed: %v", err)
+		}
+		localCh <- nil
+	}()
+
+	// Run the container side.
+	contCh := make(chan error)
+	go func() {
+		// Wait for the final statement. This structure has the side effect
+		// that all container logs will appear within the individual test
+		// context.
+		if _, err := d.WaitForOutput(ctx, TerminalStatement, TestTimeout); err != nil {
+			contCh <- fmt.Errorf("test failed: %v", err)
+		}
+		contCh <- nil
+	}()
+
+	var localDone bool
+	var contDone bool
+	for {
+		select {
+		case err := <-localCh:
+			if err != nil {
+				t.Fatal(err)
+			}
+			if test.LocalSufficient() || contDone {
+				return
+			}
+			localDone = true
+		case err := <-contCh:
+			if err != nil {
+				t.Fatal(err)
+			}
+			if test.ContainerSufficient() || localDone {
```

Comment:
nit: Are there any test cases where ContainerSufficient isn't false? I don't see any. Consider removing it and simplifying this part accordingly.

Reply:
This is now used (in the form of containerCase) in a number of tests.